### ISSUE TITLE
MaxCap / MinCap using ModifedFloat

### ISSUE THREAD

**Issue by seoyeon01** [Improx/ModifiedValues#1]
Since Max(Min)Cap only uses float values(Not ModifiedFloat), there are the following inconveniences.

`

    public ModifiedFloat HP = 10;
    public ModifiedFloat MaxHP = 100;
    
    private void Awake()
    {
        HP.MaxCap(MaxHP);
        MaxHP.BecameDirty += (o, e) => HP.SetDirty();
        // HP.UpdateEveryTime = true;
    }
    
`
Even if MaxHP changes, HP will not be updated properly.

so.. how about this?
`

	public static Modifier<float> TemplateModifiedMaxCap(ModifiedFloat modifiedValue, int priority = 0, int layer = 0, int order = DefaultOrders.Cap)
	{
		return Modifier<float>.NewFromLatest((latestValue) => Math.Min(latestValue, modifiedValue), priority, layer, order);
	}

	public Modifier<float> ModifiedMaxCap(ModifiedFloat modifiedValue, int priority = 0, int layer = 0)
	{
		var mod = TemplateModifiedMaxCap(modifiedValue, priority, layer);
		Attach(mod);
		modifiedValue.BecameDirty += (o, e) => SetDirty();
			
		return mod;
	}
		
`

**Comment by maxsamarin** [comment]
Thanks, good find. This is actually an inconvenience for all of out-of-the-box modifiers, such as Add, Mul, etc.. all of them. Your suggestion works. I'm trying to think is there an easy way to extend that to all ModifiedValue<Type> and all out-of-the-box modifier functions, without needing to duplicate all functions and make new versions of them like you have "ModifiedMaxCap" here.

Unrelated: I don't know if HP was just an example to illustrate the problem or your real use case, but I suggest that that HP would not be a ModifiedFloat but a tradidional float. Maybe MaxCap and MinCap are the only useful cases for that, but otherwise when a character would take damage, you would need to do HP.Add(-damage) each time, which would bloat the amount of attached modifiers every time. For "stats" like MaxHP I think it's more useful, but for HP or any kind of "progress" I just see it as a number rather than a ModifiedFloat.

**Comment by seoyeon01** [comment]
You are right. 
I also think it’s wrong to add damage as a modifier. 
Buffs and equipment are factors that can be attached or detached.
Damage, healing, etc. usually have an immediate effect, and there’s no need to manage those values (in most cases). 
I’ve discussed the exact same issue with someone on a similar project before.
In that case, for things like damage/healing used a method that modified the BaseValue.

An example like a Ring that increases health by 10 is a bit more complex.
It adds 10 to MaxHP and also increases HP  10
The problem arises if HP/MaxHP is (5/110) and remove the ring
According to logic, it should be (0/100) (MinCut 0), 
but most games would set it to (1/100). To implement such behavior, some extra work is likely required.

Anyway... 
Thinking about it again now, I think it’s better to just use float and handle the additional steps rather than modifying the BaseValue. I’ll give it a try with float


Thank you for making a good project. It was very helpful.
I’ve used a few projects for this purpose. Your project is not only easy to use and concise, but I also liked how it handled layers/priority/order and dependencies
I hope this project will continue to attract more people's attention and develop for a long time

**Comment by maxsamarin** [comment]
Thanks a lot for your comment! I was worried that it can be quite complex a system to do simple stuff with, but good that you got a hang of it. The point of it is to be flexible and to allow a wide range of buff/stat/equipment/perk/talent etc use cases like you described.

I will leave this issue open for now as a reminder for the problem you pointed out. It may be a common use case to want a Modifier to use a ModifiedValue as an argument rather than a hard value. I might end up making copies of each function to have a "Modified" version of it, like you suggested. Sounds like some manual work but on the other hand I might just automatically generate these versions based on the existing functions.

**Comment by maxsamarin** [comment]
Also, for the situation you describes with the ring where you want to change healrh each time MaxHealth is changed, a similar approach could be used like in Magic The Gathering. Instead of Health, you could instead track Damage, and Health would just be a getter => MaxHealth - Damage. But yeah if you don’t want to decrease health when MaxHealth drops or only prevent dying, special handling would be required.

**Comment by maxsamarin** [comment]
I now updated the repo with this new feature.

Each modifier method has a "Dynamic" version now. For example instead of calling Speed.MaxCap(MaxSpeed), call Speed.MaxCapDynamic(MaxSpeed). All of the functions have a version with Dynamic in the end of the name.

As a headsup, unrelated change: I also removed ModifiedEnum<YourEnum> class, because I realized that there was no need to limit the generic type to enums only. `ModifiedValue<T>` is now serialized, so one can just use  `ModifiedValue<T>` with any type T and it will be visible in the inspector (at least the base value).

Also noticed that methods such as Add() did not have the "order" parameter. All modified methods now have order as an optional parameter, which simply defaults to DefaultOrders.X.
